### ISSUE TITLE
No support for string data

### ISSUE THREAD

**Issue by immunda** [maraujop/requests-oauth#12]
This is what I think gldnspud's pull request was in reference to.

If you provide a string instead of a dict as the data arg to the request, then 'get_normalized_parameters()'  will cause an exception when it tries to access 'request.data.items()'; which is a string not a dict.

This behaviour is supported by the requests library: "There are many times that you want to send data that is not form-encoded. If you pass in a string instead of a dict, that data will be posted directly".

I'm not sure on the best solution to this - I'm using a solution similar to gldnspud's, but am checking the whether request.data is a basestring instance rather than checking the "Content-Type" header explicitly.

Thanks, Phil.

**Comment by maraujop** [comment]
Thanks Phil,

Now I understand what @gldnspud meant in issue #10. I've just pushed an adapted patch in `dev` branch. I agree it's better to check for `basestring` here, as it covers more use cases.

Looks like there is no other way to fix this, I think the checking is good enough.

If you guys know a public API to which against I can code a test, that would be great.

Cheers,
Miguel

**Comment by immunda** [comment]
Thanks Miguel, this looks to do the trick.

Just so you're aware, this will exclude the data from being included in the oauth signature generation. However, in the last few hours I've spent reading around this a bit more, this looks to be correct. From going right back the OAuth 1.0 spec, it looks as if the body of the request isn't included in the signature base string if the request's 'content-type' header isn't "application/x-www-form-urlencoded" (http://oauth.net/core/1.0/#sig_norm_param). This is made confusing as I was originally passing the whole JSON body as a urlencoded param key, as I've seen done in a couple of other places and what python-oauth2 seems to expect.

Anyway, Google OAuth requests without "application/x-www-form-urlencoded" content-type don't include the body in any way in their signature base string and given this is what the spec says, I'm inclined to think python-oauth2 is wrong. This does mean that instead of checking whether the data is a string or a dict, it looks compliant with the spec to just ignore it entirely from get_normalized_parameters() if the "content-type" header is not  "application/x-www-form-urlencoded". Kind of an inverse of @gldnspud 's original solution.

Sorry, I suck at Github markdown. Cheers, Phil.

**Comment by maraujop** [comment]
Ok Phil,

I've read the spec so many times I'm surprised I don't know it by heart now. Looks like you are right, 

This is basically what I'm doing. If `request.data` is a `basestring` I'm only using `request.params`, which means no POST parameters. So I'm understanding you are suggesting, that instead of checking if it's a `basestring` you think it's better to check if `Content-type` **is not** `application/x-www-form-urlencoded`, which does the same thing, but it's correct, because it follows the spec and it's less of a hack, as it looks now, right?

One more thing, what OAuth API are you using to test this against? I want to add a test and sleep better :)

Cheers,
Miguel

**Comment by immunda** [comment]
Hi Miguel,

That's right - yep. I appreciate it will except if it's provided a string with a form encoded request, but this seems consistent with the spec and requests' behaviour.

The Twitter API actually has some media uploads; for those they use `multipart/form-data` (ex: "https://dev.twitter.com/docs/api/1/post/statuses/update_with_media") which excludes the body in the signing params ("As you're using multipart in the body of the POST request, only parameters beginning with oauth_ should be part of the signature base string"); does this help for testing?

Thanks! :-)

**Comment by matthewryanscott** [comment]
Thank you, @immunda and @maraujop, for providing a "cleaner" way to
implement the change to requests-oauth that I had proposed in #10.

For what it's worth, I successfully use requests-oauth in this context:  An
API server uses https://bitbucket.org/david/django-oauth-plus/src to act as
an OAuth 1.0a provider.  A separate UI server uses
http://pypi.python.org/pypi/django-social-auth for the initial
authentication against the API server.  After that succeeds, the UI server
uses requests-oauth to make OAuth-signed calls to the API server.

I'll update my code to use the latest patch and report back on success or
failure.

On Sun, Feb 12, 2012 at 10:32, Miguel Araujo <
reply@reply.github.com

> wrote:
> 
> One more thing, what OAuth API are you using to test this against? I want
> to add a test and sleep better :)

**Comment by matthewryanscott** [comment]
Most recent patch confirmed to work in my scenario.  Thanks again!

**Comment by matthewryanscott** [comment]
I think I spoke too soon, and was perhaps testing against my app incorrectly.

The patch for this issue will pass MOST of the OAuth parameters needed as GET params when POST-ing, but not all -- in fact, the one that's missing is the `oauth_signature`.

However, what I -can- confirm is that by using `header_auth=True`, the `django-oauth-plus` code will correctly interpret the request.

So perhaps an appropriate way to handle this particular patch would be to 1) roll it back, or 2) fix it to ensure that `oauth_signature` is being sent.

I'm sorry I can't provide more time at the moment to contribute a fix, but thank you for your attention to this, and thank you @maraujop for recommending the `header_auth` that's available in the `dev` branch.

**Comment by maraujop** [comment]
You are right Matthew. `oauth_signature` wasn't being sent. I've fixed that.

Finally @immunda I'm using a mixture of both solutions, if you look at latest version in `dev` branch. This is because when data is a string `Content-Type` header is not set. This way we follow the spec and work with special requests cases.

Also for getting, latest requests-oauth to work, right now you will need to use my fork of the project. I've submitted a pull request that fixes and issue with hooks. Specifically this one:
https://github.com/kennethreitz/requests/issues/445

One more thing, I've added a test using Twitter API as suggested by @immunda (Phil) using `multipart/form-data` and it's working correctly. So I'm quite confident this one is finally fixed.

Cheers,
Miguel
